### PR TITLE
Assessment Tool Groups UI on Categorization Page

### LINKED ISSUE #118: I can group assessment tool columns together
When I categorize columns, I want to be able to indicate that multiple columns belong to the same assessment tool (e.g. subscales of the same tool), so that
- I don't have to label every column individually with the same concept
- I can indicate the columns that are not allowed to have missing values for this tool to be considered available for a subject

Done when:

- [ ] I can create a new tool group with the user interface on the categorization page
- [ ] A new tool group gets a default name (e.g. tool 1) if I don't provide one
- [ ] Adding columns to tool groups works similarly to mapping columns to age, sex, and diagnosis. I can click on a tool group (or make a new one), and then click on a column to assign it to this tool group
- [ ] I can remove columns from a tool group via the categorization page UI
- [ ] A column can only be added to a single tool group.

### PR DESCRIPTION
**Pull request by jarmoza**
This branch introduces the user interface on the categorization page for grouping together table columns that have been categorized as assessment tools. (The interface consists of instructions text, a textbox, multi-selectbox, button, and table that sits beneath the categorization interface.)

Expected behaviors:
- [ ] User can type in a group name in a textbox in the tool group interface. Any name is acceptable as long as it has not yet been added as tool group in previous grouping action
- [ ] As the user categorizes columns as an 'Assessment Tool', column names populate a multiselect listbox in the tool group interface at the bottom of the screen
- [ ] The multiselect box adds columns to a group via click, or using 'shift' or 'ctrl' keys to select multiple items (as one would in a file system)
- [ ] Clicking the `+ Create Tool Group` button adds the tool group with its name and selected columns to a table on the right of the tool group interface.
- [ ] Once columns have been added to the tool group table they are grayed out and no longer selectable
- [ ] The `+ Create Tool Group` button is only enabled for clicking when a unique name has been entered into the textbox and at least one column is selected in the multi-selectbox.
- [ ] If the user attempts to enter non-unique name, text indicating this name is already used will be displayed at the bottom of the tool group interface (`Group already exists in table`)
- [ ] Tool groups can be removed from the table by clicking the `X Remove Tool Group` button in the last column of its table row. The table row is deleted and the columns in that now-deleted group are ungrayed in the multi-select box and are available for adding to a new tool group.

Resolves #118

### CODE REVIEW

File: components/categ-tool-group.vue [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+                head-variant="dark"
+                :items="assessmentToolGroups.items"
+                :fields="assessmentToolGroups.fields">
+
+                <!-- Blank row for empty table for aesthetics -->
+                <template #top-row v-if="hasNoGroups">
+                    <!-- Adding &nbsp; to the cell so that it maintains the standard cell height -->
+                    <td
+                        v-for="index in assessmentToolGroups.fields.length"
+                        :key="assessmentToolGroups.fields[index]">
```
Comment (surchs) [inline components/categ-tool-group.vue:79]: Did you mean to add `key` here? I get a bunch of warnings about non-literal keys and duplicate keys:

image

Comment (surchs) [reply components/categ-tool-group.vue:79]: To clarify, I mean: `assessmentToolGroups.fields[index].key`?

**Review by surchs — CHANGES_REQUESTED** [review]
Thanks for the PR @jarmoza. This is a cool user interface and I really like the tool group idea! I do have some comments:

## User interface / experience
I really like the idea of adding a new tool group and then adding to it. What I would suggest is to change the user experience slightly: rather than a two-step process where I first label all of the "assessment tool" columns, and then sort them to their respective tools in a second step, I would roll the whole thing into one step like so:

1. Click on a button (+) to add a new tool group (e.g. directly below the category table). This can either ask for a tool name or apply a default name (e.g. `tool 1`). 
2. Otherwise this tool group behaves like another category. I.e.: it gets a tab in the annotation table that can be clicked. And: I can now go and assign columns to this tool group just like I do with other categories (e.g. also to undo). 
3. The tool gets one extra UI elements: little delete button (e.g. a small X right-aligned in the tab) to remove it again when clicked

Three main reasons for this:
1. We have only one user pattern (select left, assign right) with fewer steps to explain and keep track of
2. The state is clear. Right now, I can label a bunch of columns as "assessment tools", create a new tool group, assign them to it, and then go back and unlabel them or label the columns as something other than "assessment tools". So now I have a confusing state where my columns are both part of an assessment tool group and also mapped to e.g. "age".
3. There is no confusion about what to do with columns that have been labeled as "assessment tool" but not assigned to any tool group

## Overall
- this doesn't seem to be hooked up to the global store yet. I think it makes sense to add that to this PR so that any future work can use this (e.g. if we want to work in parallel on something that requires the tool-group info). 

## Code
- since this branch is currently behind master, you don't have the current linter rules. There are several lines with trailing whitespace, best to remove these now before we encounter them again. As you merge master into this branch (or rebase), please also run the linter with the --fix flag to get rid of these
- there are some assignments (I marked them) that, as far as I know, are not reactive in Vue (e.g. adding a new property to an object with `my_object[my_prop] = someval;`). This seems to not be reflected in the app behaviour (e.g. these changes seem to always trigger reactive renders), but unless there is a good reason to not use [the recommended assignment patterns from Vue](https://v2.vuejs.org/v2/guide/reactivity.html), you should refactor these parts.

File: components/categ-tool-group.vue [hunk 2]
```diff
@@ -0,0 +1,261 @@
+<template>
```
[the remark this answers is not in the document]
Comment (jarmoza) [reply components/categ-tool-group.vue:1]: Agreed. I was on the verge of a name change. Thinking about prepending all subcomponents of the categorization page with `categ`, but `category` works too.

File: components/categ-tool-group.vue [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                    createToolGroupButton: "+ Create Tool Group",
+                    instructions: "Choose a name and columns and then click the 'create' button",
+                    removeToolGroupButton: "x Remove Tool Group",
+                    title: "Assessment Tool Groups",
+                    toolGroupNamePlaceholder: "Type name of assessment tool group here..."
+                }
+            };
+        },
+
+        computed: {
```
[the remark this answers is not in the document]
Comment (jarmoza) [reply components/categ-tool-group.vue:158]: Thanks!

File: components/categ-tool-group.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (jarmoza) [reply components/categ-tool-group.vue:1]: I'll address this in the next PR.

File: components/categ-tool-group.vue (hunk 1 above)
Comment (jarmoza) [reply components/categ-tool-group.vue:79]: I caught this one in my current work. Didn't see the warnings. The fix is

```
                        v-for="field in assessmentToolGroups.fields"
                        :key="field.key">
```

File: components/categ-tool-group.vue [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            },
+
+            removeToolGroup(p_row) {
+
+                // 1. Clear this group's row in the table
+                let groupIndex = this.assessmentToolGroups.items.findIndex(x => x.name === p_row.item.name);
+                this.assessmentToolGroups.items.splice(groupIndex, 1);
+
+                // 2. Remove this group from the tool group map
+                delete this.toolGroups[p_row.item.name];
```
[the remark this answers is not in the document]
Comment (jarmoza) [reply components/categ-tool-group.vue:247]: Fixed. Will have to see how this is done in a store mutation in the next PR.

**Comment by jarmoza** [comment]
>this doesn't seem to be hooked up to the global store yet. I think it makes sense to add that to this PR so that any future work can use this (e.g. if we want to work in parallel on something that requires the tool-group info).

Store hookup is addressed in the very next PR for this week's sprint (wherein the tool groups start to have to be used outside of the page). Also addressed in the next PR:

1. Tool groups properly repopulating when the categorization page is returned to
2. Access to the annotation page is contingent on all categorized tools being grouped
3. Access to the annotation page is disabled in the scenario where tool groups are removed from the table

With that in mind, I'd like to merge this PR and followup with the (more) fully functional tool grouping feature right after this.

**Comment by surchs** [comment]
@jarmoza: check out this tutorial to resolve your merge conflict: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow There is a section called "Example".

You'll  have to merge master into your feature branch

**Comment by surchs** [comment]
> With that in mind, I'd like to merge this PR and followup with the (more) fully functional tool grouping feature right after this.
OK, sounds good

> I did not however detect any linter whitespace errors.
You will only see these once you have merged (or rebased) the current master branch onto this one because the linter rules have changed since you made this fork. The main difference is that trailing whitespace is now disallowed everywhere.

**Comment by jarmoza** [comment]
> * there are some assignments (I marked them) that, as far as I know, are not reactive in Vue (e.g. adding a new property to an object with `my_object[my_prop] = someval;`). This seems to not be reflected in the app behaviour (e.g. these changes seem to always trigger reactive renders), but unless there is a good reason to not use [the recommended assignment patterns from Vue](https://v2.vuejs.org/v2/guide/reactivity.html), you should refactor these parts.

An important comment on this part. Though I made the changes for this PR, it turns out that once these changes are made in the store the regular JS assign and delete functionality is just fine, even necessary since Vue.delete and Vue.set are not available. The Vuex store is reactive: https://vuex.vuejs.org/guide/

>Vuex stores are reactive. When Vue components retrieve state from it, they will reactively and efficiently update if the store's state changes.

**Pull request merged**